File: compiler/src/lexer.ts
export type TokenType =
  | 'ident'
  | 'string'
  | 'number'
  | 'punct'
  | 'eof';

export interface Token {
  type: TokenType;
  text: string;
  line: number;
  col: number;
}

export class LexError extends Error {
  constructor(message: string, public line: number, public col: number) {
    super(`${message} (line ${line}, col ${col})`);
  }
}

const PUNCTUATION_TWO_CHAR = ['==', '!=', '<=', '>='];
const PUNCTUATION_ONE_CHAR = ['{', '}', '(', ')', ',', '.', ':', '?', '=', '+', '-', '<', '>'];

export function tokenize(source: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;
  let line = 1;
  let col = 1;

  function advance(n = 1) {
    for (let k = 0; k < n; k++) {
      if (source[i] === '\n') {
        line++;
        col = 1;
      } else {
        col++;
      }
      i++;
    }
  }

  while (i < source.length) {
    const ch = source[i];

    // whitespace
    if (ch === ' ' || ch === '\t' || ch === '\r' || ch === '\n') {
      advance();
      continue;
    }

    // line comment
    if (ch === '/' && source[i + 1] === '/') {
      while (i < source.length && source[i] !== '\n') advance();
      continue;
    }

    const startLine = line;
    const startCol = col;

    // string literal
    if (ch === '"') {
      let value = '';
      advance();
      while (i < source.length && source[i] !== '"') {
        if (source[i] === '\n') {
          throw new LexError('unterminated string literal', startLine, startCol);
        }
        value += source[i];
        advance();
      }
      if (i >= source.length) {
        throw new LexError('unterminated string literal', startLine, startCol);
      }
      advance(); // closing quote
      tokens.push({ type: 'string', text: value, line: startLine, col: startCol });
      continue;
    }

    // number literal (integer or decimal)
    if (/[0-9]/.test(ch)) {
      let text = '';
      while (i < source.length && /[0-9.]/.test(source[i])) {
        text += source[i];
        advance();
      }
      tokens.push({ type: 'number', text, line: startLine, col: startCol });
      continue;
    }

    // identifier / keyword
    if (/[A-Za-z_]/.test(ch)) {
      let text = '';
      while (i < source.length && /[A-Za-z0-9_]/.test(source[i])) {
        text += source[i];
        advance();
      }
      tokens.push({ type: 'ident', text, line: startLine, col: startCol });
      continue;
    }

    // two-char punctuation
    const two = source.slice(i, i + 2);
    if (PUNCTUATION_TWO_CHAR.includes(two)) {
      advance(2);
      tokens.push({ type: 'punct', text: two, line: startLine, col: startCol });
      continue;
    }

    // one-char punctuation
    if (PUNCTUATION_ONE_CHAR.includes(ch)) {
      advance();
      tokens.push({ type: 'punct', text: ch, line: startLine, col: startCol });
      continue;
    }

    throw new LexError(`unexpected character '${ch}'`, startLine, startCol);
  }

  tokens.push({ type: 'eof', text: '', line, col });
  return tokens;
}

File: compiler/src/main/kotlin/velle/Lexer.kt
package velle

/**
 * Lexer for Velle (grammar.md, "Lexical layer").
 *
 * Line-orientation: a newline is a token (statement / member separator), except
 * inside parentheses, where newlines are insignificant and suppressed here.
 * Comments run from `--` to end of line. Text literals are single-line with
 * Kotlin-style escapes minus `\$`.
 */
class Lexer(private val src: String) {

    private var pos = 0
    private var line = 1
    private var col = 1
    private var parenDepth = 0
    private val tokens = mutableListOf<Token>()

    fun lex(): List<Token> {
        while (pos < src.length) {
            val c = src[pos]
            when {
                c == '\n' -> {
                    if (parenDepth == 0) emit(TokType.NEWLINE, "\n")
                    advance()
                    line++; col = 1
                }
                c == ' ' || c == '\t' || c == '\r' -> advance()
                c == '-' && peek(1) == '-' -> skipComment()
                c.isDigit() -> number()
                c == '"' -> textLiteral()
                c.isLetter() -> identifier()
                else -> symbol()
            }
        }
        emit(TokType.EOF, "")
        return tokens
    }

    private fun skipComment() {
        while (pos < src.length && src[pos] != '\n') advance()
    }

    private fun number() {
        val startCol = col
        val sb = StringBuilder()
        while (pos < src.length && src[pos].isDigit()) { sb.append(src[pos]); advance() }
        if (pos < src.length && src[pos] == '.' && peek(1)?.isDigit() == true) {
            sb.append('.'); advance()
            while (pos < src.length && src[pos].isDigit()) { sb.append(src[pos]); advance() }
            tokens.add(Token(TokType.DEC, sb.toString(), line, startCol))
        } else {
            tokens.add(Token(TokType.INT, sb.toString(), line, startCol))
        }
    }

    private fun textLiteral() {
        val startLine = line
        val startCol = col
        advance() // opening quote
        val sb = StringBuilder()
        while (true) {
            if (pos >= src.length || src[pos] == '\n')
                throw VelleSyntaxError("unterminated text literal", startLine, startCol)
            val c = src[pos]
            if (c == '"') { advance(); break }
            if (c == '\\') {
                advance()
                if (pos >= src.length) throw VelleSyntaxError("unterminated escape", line, col)
                when (val e = src[pos]) {
                    '"' -> sb.append('"')
                    '\\' -> sb.append('\\')
                    'n' -> sb.append('\n')
                    't' -> sb.append('\t')
                    'r' -> sb.append('\r')
                    'u' -> {
                        val hex = src.substring(pos + 1, minOf(pos + 5, src.length))
                        if (hex.length < 4) throw VelleSyntaxError("bad \\u escape", line, col)
                        sb.append(hex.toInt(16).toChar())
                        repeat(4) { advance() }
                    }
                    else -> throw VelleSyntaxError("unknown escape \\$e", line, col)
                }
                advance()
            } else {
                sb.append(c); advance()
            }
        }
        tokens.add(Token(TokType.TEXT, sb.toString(), startLine, startCol))
    }

    private fun identifier() {
        val startCol = col
        val sb = StringBuilder()
        while (pos < src.length && (src[pos].isLetterOrDigit())) { sb.append(src[pos]); advance() }
        val word = sb.toString()
        val type = when {
            word in KEYWORDS -> TokType.KW
            word[0].isUpperCase() -> TokType.UIDENT
            else -> TokType.LIDENT
        }
        tokens.add(Token(type, word, line, startCol))
    }

    private fun symbol() {
        val c = src[pos]
        val n = peek(1)
        fun two(t: TokType, text: String) { emit(t, text); advance(); advance() }
        fun one(t: TokType) { emit(t, c.toString()); advance() }
        when {
            c == '=' && n == '=' -> two(TokType.EQ, "==")
            c == '=' -> one(TokType.ASSIGN)
            c == '!' && n == '=' -> two(TokType.NEQ, "!=")
            c == '<' && n == '=' -> two(TokType.LE, "<=")
            c == '<' -> one(TokType.LT)
            c == '>' && n == '=' -> two(TokType.GE, ">=")
            c == '>' -> one(TokType.GT)
            c == '?' && n == '.' -> two(TokType.QDOT, "?.")
            c == '?' -> one(TokType.QMARK)
            c == '.' -> one(TokType.DOT)
            c == '+' -> one(TokType.PLUS)
            c == '-' -> one(TokType.MINUS)
            c == '*' -> one(TokType.STAR)
            c == '/' -> one(TokType.SLASH)
            c == ':' -> one(TokType.COLON)
            c == ',' -> one(TokType.COMMA)
            c == '{' -> one(TokType.LBRACE)
            c == '}' -> one(TokType.RBRACE)
            c == '(' -> { parenDepth++; one(TokType.LPAREN) }
            c == ')' -> { parenDepth--; one(TokType.RPAREN) }
            else -> throw VelleSyntaxError("unexpected character '$c'", line, col)
        }
    }

    private fun emit(type: TokType, text: String) = tokens.add(Token(type, text, line, col))
    private fun advance() { pos++; col++ }
    private fun peek(ahead: Int): Char? = src.getOrNull(pos + ahead)
}
